Project2015To2017/Reading/ProjectReader.cs

```diff
@@ -1,28 +1,48 @@
+using Project2015To2017.Definition;
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.IO;
 using System.Linq;
 using System.Xml;
 using System.Xml.Linq;
-using Project2015To2017.Definition;
 using static Project2015To2017.Definition.Project;
 
 namespace Project2015To2017.Reading
 {
 	public class ProjectReader
 	{
-		public Project Read(string filePath)
+		public static readonly ProjectReader Instance = new ProjectReader();
+		public bool EnableCaching { get; set; }
+
+		public Project Read(string filePath, IProgress<string> progress = null)
 		{
-			return Read(filePath, new Progress<string>(_ => { }));
+			progress = progress ?? new Progress<string>(_ => { });
+			return Read(new FileInfo(filePath), progress);
 		}
 
-		public Project Read(string filePath, IProgress<string> progress)
+		/// <summary>
+		/// Process-lifetime-long cache of loaded projects
+		/// <remarks>
+		/// It was supposed to be of <see cref="System.Runtime.Caching.ObjectCache"/> type, but that can still remove items from cache under memory pressure,
+		/// while we still need availability guarantees.
+		/// </remarks>
+		/// </summary>
+		private readonly Dictionary<int, Project> _cache = new Dictionary<int, Project>();
+
+		public Project Read(FileInfo file, IProgress<string> progress, bool loadModern = false)
 		{
+			var filePath = file.FullName;
+			var hash = filePath.GetHashCode();
+			if (EnableCaching && _cache.TryGetValue(hash, out var projectDefinition))
+			{
+				return projectDefinition;
+			}
+
 			XDocument projectXml;
 			using (var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
 			{
-				projectXml = XDocument.Load(stream);
+				projectXml = XDocument.Load(stream, LoadOptions.SetLineInfo);
```

Comment:
To be able later to find line number for XML element data source stored in new fields in a couple of classes.